src/merkle/new_smt/mod.rs

```diff
@@ -0,0 +1,272 @@
+            }
+        }
+    }
+
+    fn get_leaf(&self, key: &Self::Key) -> Self::Leaf {
+        let leaf_pos = LeafIndex::<NEW_SMT_DEPTH>::from(*key).value();
+
+        match self.leaves.get(&leaf_pos) {
+            Some(leaf) => leaf.clone(),
+            None => NewSmtLeaf::Single((
+                *key,
+                Word::from(*EmptySubtreeRoots::entry(self.depth(), self.depth())),
+            )),
+        }
+    }
+
+    fn hash_leaf(leaf: &Self::Leaf) -> RpoDigest {
+        leaf.hash()
+    }
+}
+
+impl Default for NewSmt {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+// KEY
+// ================================================================================================
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
+pub struct NewSmtKey {
+    word: Word,
+}
+
+impl From<NewSmtKey> for LeafIndex<NEW_SMT_DEPTH> {
+    fn from(key: NewSmtKey) -> Self {
+        let most_significant_felt = key.word[0];
+        Self::new_max_depth(most_significant_felt.as_int())
+    }
+}
+
+impl Ord for NewSmtKey {
+    fn cmp(&self, other: &Self) -> Ordering {
+        // Note that the indices are reversed. This is because a `Word` is treated as little-endian
+        // (i.e. most significant Felt is at index 3). In comparing integer arrays, we want to
+        // compare the most significant felts first, and so on until the least signifcant felt.
+        let self_word: [u64; 4] = [
+            self.word[3].as_int(),
+            self.word[2].as_int(),
+            self.word[1].as_int(),
+            self.word[0].as_int(),
+        ];
+        let other_word: [u64; 4] = [
+            other.word[3].as_int(),
+            other.word[2].as_int(),
+            other.word[1].as_int(),
+            other.word[0].as_int(),
+        ];
+
+        self_word.cmp(&other_word)
+    }
+}
+
+impl PartialOrd for NewSmtKey {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+// LEAF
+// ================================================================================================
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize, serde::Serialize))]
+pub enum NewSmtLeaf {
+    Single((NewSmtKey, Word)),
+    Multiple(Vec<(NewSmtKey, Word)>),
+}
+
+impl NewSmtLeaf {
+    pub fn hash(&self) -> RpoDigest {
+        fn kv_to_elements((key, value): &(NewSmtKey, Word)) -> impl Iterator<Item = Felt> + '_ {
+            let key_elements = key.word.iter().copied();
+            let value_elements = value.iter().copied();
+
+            key_elements.chain(value_elements)
+        }
+
+        let elements: Vec<Felt> = match self {
+            NewSmtLeaf::Single(kv) => kv_to_elements(kv).collect(),
+            NewSmtLeaf::Multiple(kvs) => kvs.iter().flat_map(kv_to_elements).collect(),
+        };
+
+        Rpo256::hash_elements(&elements)
+    }
```

Comment:
I would probably handle the case of single vs. multiple entries separately here. The main reason is that for the single entry case (which is going to be by far the most dominant), we can just use `Rpo256::merge()` functions which would not require allocating vectors.

Another option is to transmute `(NewSmtKey, Word) -> (Word, Word) -> &[Felt]` - and then we can apply `hash_elements()` in both cases without allocating vectors - but I'm not sure it's worth it.

Reply:
Done